internal/nri/plugin.go

```diff
@@ -0,0 +1,385 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to get NUMA node for device %s: %w", uuid, err)
+		}
+		numaNodes = append(numaNodes, node)
+	}
+
+	// Convert to strings
+	nodeStrs := make([]string, len(numaNodes))
+	for i, node := range numaNodes {
+		nodeStrs[i] = strconv.Itoa(node)
+	}
+
+	return nodeStrs, nil
+}
+
+// getAssignedDevices returns the list of GPU and MIG device UUIDs assigned to a container.
+// The function should branch based on the configuration in p.devicePluginConfig to determine how to
+// handle different device types (e.g., MIG devices, replicated resources) and their
+// NUMA node assignments.
+func (p *NUMAFilterPlugin) getAssignedDevices(c *api.Container) ([]string, error) {
+	// TODO: Implement branching logic based on p.devicePluginConfig to handle
+	// different device list strategies
+
+	// Look for NVIDIA_VISIBLE_DEVICES environment variable
+	for _, env := range c.Env {
+		if strings.HasPrefix(env, "NVIDIA_VISIBLE_DEVICES=") {
+			// Get the value after the equals sign
+			value := strings.TrimPrefix(env, "NVIDIA_VISIBLE_DEVICES=")
+
+			// Handle special cases
+			switch value {
+			case "", "void":
+				return nil, nil
+			case "all":
+				uuids, err := p.getAllDeviceUUIDs()
+				if err != nil {
+					return nil, fmt.Errorf("failed to get all device UUIDs: %w", err)
+				}
+				return uuids, nil
+			default:
+				// Handle comma-separated list of UUIDs
+				return strings.Split(value, ","), nil
+			}
+		}
+	}
+	return nil, nil
+}
+
+// getDeviceNUMANode gets the NUMA node for a device by UUID, first checking the cache
+// and then looking it up if not found.
+func (p *NUMAFilterPlugin) getDeviceNUMANode(uuid string) (int, error) {
+	// First check the cache
+	if node, ok := p.cache.deviceNodes[uuid]; ok {
+		return node, nil
+	}
+
+	// If not in cache, try to look it up
+	device, ret := p.nvml.DeviceGetHandleByUUID(uuid)
+	if ret != nvml.SUCCESS {
+		return 0, fmt.Errorf("failed to get device handle for UUID %s: %v", uuid, ret)
+	}
+
+	// Get NUMA node
+	node, ret := device.GetNumaNodeId()
+	if ret != nvml.SUCCESS {
+		return 0, fmt.Errorf("failed to get NUMA node for device %s: %v", uuid, ret)
+	}
+
+	// Cache the result
+	p.cache.deviceNodes[uuid] = node
+	return node, nil
+}
+
+// getAllSystemNUMANodes gets all NUMA nodes in the system that have CPUs assigned to them
+func (p *NUMAFilterPlugin) getAllSystemNUMANodes() ([]int, error) {
+	// Read /sys/devices/system/node/node* to get NUMA nodes
+	nodes, err := filepath.Glob("/sys/devices/system/node/node*")
+	if err != nil {
+		return nil, fmt.Errorf("failed to list NUMA nodes: %w", err)
+	}
+
+	systemNodes := make([]int, 0, len(nodes))
+	for _, node := range nodes {
+		nodeID := filepath.Base(node)[4:] // Remove "node" prefix
+		// Check if the node has any CPUs assigned
+		cpulist, err := os.ReadFile(filepath.Join(node, "cpulist"))
+		if err != nil {
+			continue
+		}
+
+		// Only include nodes with non-empty CPU lists
```

Comment:
So, we define a "system node" to be any NUMA node that has a CPU assigned. Is that right? Is that the one, definite criterion or is there any other criterion? Should we call them `cpuNodes`? Is there a better way to detect system nodes?

Reply:
That is the only criteria. We usually refer to this kind of memory as "system memory", which is why I picked "system nodes", but I'm open to suggestions.